bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1982,26 +1982,43 @@ private static boolean checkIsLikeTupleType(Object sourceValue, BTupleType targe
         }
 
         ArrayValue source = (ArrayValue) sourceValue;
-        if (source.size() != targetType.getTupleTypes().size()) {
+        List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        int sourceTypeSize = source.size();
+        int targetTypeSize = targetTypes.size();
+        BType targetRestType = targetType.getRestType();
+
+        if (targetRestType == null && sourceTypeSize != targetTypeSize) {
+            return false;
+        }
+        if (targetRestType != null && sourceTypeSize <= targetTypeSize) {
             return false;
         }
 
-        int bound = source.size();
-        for (int i = 0; i < bound; i++) {
-            BType elementType = getArrayElementType(source, i);
-            if (BTypes.isValueType(elementType)) {
-                if (!checkIsType(elementType, targetType.getTupleTypes().get(i), new ArrayList<>())) {
-                    return false;
-                }
-            } else {
-                if (!checkIsLikeType(source.getRefValue(i), targetType.getTupleTypes().get(i), unresolvedValues,
-                        allowNumericConversion)) {
-                    return false;
-                }
+        for (int i = 0; i < targetTypeSize; i++) {
+            if (!checkTupleElementType(getArrayElementType(source, i), targetTypes.get(i), source.getRefValue(i),
+                    unresolvedValues, allowNumericConversion)) {
+                return false;
+            }
+        }
+
+        for (int i = targetTypeSize; i < sourceTypeSize; i++) {
+            if (!checkTupleElementType(getArrayElementType(source, i), targetRestType, source.getRefValue(i),
+                    unresolvedValues, allowNumericConversion)) {
+                return false;
             }
         }
         return true;
+    }
 
+    private static boolean checkTupleElementType(BType sourceElementType, BType targetElementType,
+                                                 Object sourceRefValue, List<TypeValuePair> unresolvedValues,
+                                                 boolean allowNumericConversion) {
+        if (BTypes.isValueType(sourceElementType)) {
```

Comment:
There's also an issue with this check that results in calling `checkIsType`.

The following panics even though it shouldn't because the immutable list `[1, 2]` belongs to `[byte, byte]`.

```ballerina
public function main() {
    [int, int] x = [1, 2];
    any a = x.cloneReadOnly();

    if !(a is [byte, byte]) {
        panic error("didn't match");
    }
}
```

Reply:
Fixed.